### ISSUE TITLE
[QUESTION]Emission and convection code sequence 

### ISSUE THREAD

**Issue by huoxiao-C** [geoschem/geos-chem#125]
Dear development team:
     I am developing the geos-chem co2 adjoint code based on geos-chem 12.4.0 version. Because The adjoint code is related with the geos-chem code, therefore I would like to exchange the code order below, I don't know whether it will cause error results .
    The following code exists in main.F, 
```  
         IF ( ITS_TIME_FOR_EMIS() ) THEN      //**_before exchange start _**
#if defined( BPCH_DIAG )
            ! Increment emission counter
            CALL Set_Ct_Emis( INCREMENT=.TRUE. )
#endif
            !===========================================================
            !         ***** D R Y   D E P O S I T I O N *****
            !===========================================================
            IF ( LDRYD ) THEN
#if defined( USE_TIMERS )
               CALL GEOS_Timer_Start( "Dry deposition", RC )
#endif
               ! Compute drydep velocities
               CALL Do_Drydep( am_I_Root,  Input_Opt,  State_Chm, 
     &                         State_Diag, State_Grid, State_Met, RC )
               ! Trap potential errors
               IF ( RC /= GC_SUCCESS ) THEN
                  ErrMsg = 'Error encountered in "Do_Drydep!"!'
                  CALL Error_Stop( ErrMsg, ThisLoc )
               ENDIF
#if defined( USE_TIMERS )
               CALL GEOS_Timer_End ( "Dry deposition", RC )
#endif
               !### Debug
               IF ( prtDebug ) THEN 
                  CALL Debug_Msg( '### MAIN: a DO_DRYDEP' )
               ENDIF
            ENDIF
 #if defined( USE_TIMERS )
            CALL GEOS_Timer_Start( "HEMCO", RC )
#endif          
            ! Is it time for emissions?
            TimeForEmis = ITS_TIME_FOR_EMIS()
            ! EMISSIONS_RUN will call HEMCO run phase 2. HEMCO run phase
            ! only calculates emissions. All data has been read to disk
            ! in phase 1 at the beginning of the time step. 
            ! (ckeller, 4/1/15)
            CALL Emissions_Run( am_I_Root,  Input_Opt,  State_Chm, 
     &                          State_Diag, State_Grid, State_Met,
     &                          TimeForEmis, 2,         RC )
            ! Trap potential errors
            IF ( RC /= GC_SUCCESS ) THEN
               ErrMsg = 
     &           'Error encountered in "Emissions_Run"! after drydep!'
               CALL Error_Stop( ErrMsg, ThisLoc )
            ENDIF
            !### Debug
            IF ( prtDebug ) THEN 
               CALL Debug_Msg( '### MAIN: a HEMCO PHASE 2' )
            ENDIF
#if defined( USE_TIMERS )
            CALL GEOS_Timer_End( "HEMCO", RC )
#endif
         ENDIF                                //**_before exchange end_**
         !-------------------------------
         ! Test for convection timestep
         !-------------------------------
         IF ( ITS_TIME_FOR_CONV() ) THEN          
#if defined( BPCH_DIAG )
            ! Increment the convection timestep
            CALL Set_Ct_Conv( INCREMENT=.TRUE. )
#endif
            !===========================================================
            !      ***** M I X E D   L A Y E R   M I X I N G *****
            !===========================================================
#if defined( USE_TIMERS )
            CALL GEOS_Timer_Start( "Boundary layer mixing", RC )
#endif
            ! Note: mixing routine expects tracers in v/v
            ! DO_MIXING applies the tracer tendencies (dry deposition,
            ! emission rates) to the tracer arrays and performs PBL
            ! mixing. 
            ! In the non-local PBL scheme, dry deposition and emission
            ! fluxes below the PBL are handled within the PBL mixing
            ! routine. Otherwise, tracer concentrations are first updated
            ! and the full-mixing is then applied.
            ! (ckeller, 3/5/15)
            ! NOTE: Tracer concentration units are converted locally
            ! to [v/v dry air] for mixing. Eventually mixing should
            ! be updated to use [kg/kg total air] (ewl, 9/18/15)
            CALL Do_Mixing( am_I_Root,  Input_Opt,  State_Chm,
     &                      State_Diag, State_Grid, State_Met, RC )
            ! Trap potential errors
            IF ( RC /= GC_SUCCESS ) THEN
               ErrMsg = 'Error encountered in "Do_Mixing"!'
               CALL ERror_Stop( ErrMsg, ThisLoc )
            ENDIF
#if defined( USE_TIMERS )
            CALL GEOS_Timer_End( "Boundary layer mixing", RC )
#endif
            IF ( prtDebug ) CALL Debug_Msg( '### MAIN: a TURBDAY:2' )
            !===========================================================
            !        ***** C L O U D   C O N V E C T I O N *****
            !===========================================================
            IF ( LCONV ) THEN
#if defined( USE_TIMERS )
               CALL GEOS_Timer_Start( "Convection", RC )
#endif
               ! Call the appropriate convection routine
               ! NOTE: Tracer concentration units are converted locally
               ! to [kg/kg total air] for convection (ewl, 9/18/15)
               CALL Do_Convection( am_I_Root,  Input_Opt,  State_Chm,
     &                             State_Diag, State_Grid, State_Met,
     &                             RC )
               ! Trap potential errors
               IF ( RC /= GC_SUCCESS ) THEN
                  ErrMsg = 'Error encountered in "Do_Convection"!'
                  CALL Error_Stop( ErrMsg, ThisLoc )
               ENDIF
               !### Debug
               IF ( prtDebug ) THEN
                  CALL Debug_Msg( '### MAIN: a CONVECTION' )
               ENDIF
#if defined( USE_TIMERS )
               CALL GEOS_Timer_End( "Convection", RC )
#endif
            ENDIF 
         ENDIF
 ```
 I want the emission code segment to be behind the convection code segment, the result is below:
```
          !===========================================================
           !      ***** M I X E D   L A Y E R   M I X I N G *****
           !===========================================================
#if defined( USE_TIMERS )
            CALL GEOS_Timer_Start( "Boundary layer mixing", RC )
#endif
            ! Note: mixing routine expects tracers in v/v
            ! DO_MIXING applies the tracer tendencies (dry deposition,
            ! emission rates) to the tracer arrays and performs PBL
            ! mixing. 
            ! In the non-local PBL scheme, dry deposition and emission
            ! fluxes below the PBL are handled within the PBL mixing
            ! routine. Otherwise, tracer concentrations are first updated
            ! and the full-mixing is then applied.
            ! (ckeller, 3/5/15)
            ! NOTE: Tracer concentration units are converted locally
            ! to [v/v dry air] for mixing. Eventually mixing should
            ! be updated to use [kg/kg total air] (ewl, 9/18/15)
            CALL Do_Mixing( am_I_Root,  Input_Opt,  State_Chm,
     &                      State_Diag, State_Grid, State_Met, RC )
            ! Trap potential errors
            IF ( RC /= GC_SUCCESS ) THEN
               ErrMsg = 'Error encountered in "Do_Mixing"!'
               CALL ERror_Stop( ErrMsg, ThisLoc )
            ENDIF
#if defined( USE_TIMERS )
            CALL GEOS_Timer_End( "Boundary layer mixing", RC )
#endif
            IF ( prtDebug ) CALL Debug_Msg( '### MAIN: a TURBDAY:2' )
            !===========================================================
            !        ***** C L O U D   C O N V E C T I O N *****
            !===========================================================
            IF ( LCONV ) THEN
#if defined( USE_TIMERS )
               CALL GEOS_Timer_Start( "Convection", RC )
#endif
               ! Call the appropriate convection routine
               ! NOTE: Tracer concentration units are converted locally
               ! to [kg/kg total air] for convection (ewl, 9/18/15)
               CALL Do_Convection( am_I_Root,  Input_Opt,  State_Chm,
     &                             State_Diag, State_Grid, State_Met,
     &                             RC )
               ! Trap potential errors
               IF ( RC /= GC_SUCCESS ) THEN
                  ErrMsg = 'Error encountered in "Do_Convection"!'
                  CALL Error_Stop( ErrMsg, ThisLoc )
               ENDIF
               !### Debug
               IF ( prtDebug ) THEN
                  CALL Debug_Msg( '### MAIN: a CONVECTION' )
               ENDIF
#if defined( USE_TIMERS )
               CALL GEOS_Timer_End( "Convection", RC )
#endif
            ENDIF 
         ENDIF       
         IF ( ITS_TIME_FOR_EMIS() ) THEN           //**_after exchange start _**
#if defined( BPCH_DIAG )
            ! Increment emission counter
            CALL Set_Ct_Emis( INCREMENT=.TRUE. )
#endif
            !===========================================================
            !         ***** D R Y   D E P O S I T I O N *****
            !===========================================================
            IF ( LDRYD ) THEN
#if defined( USE_TIMERS )
               CALL GEOS_Timer_Start( "Dry deposition", RC )
#endif
               ! Compute drydep velocities
               CALL Do_Drydep( am_I_Root,  Input_Opt,  State_Chm, 
     &                         State_Diag, State_Grid, State_Met, RC )
               ! Trap potential errors
               IF ( RC /= GC_SUCCESS ) THEN
                  ErrMsg = 'Error encountered in "Do_Drydep!"!'
                  CALL Error_Stop( ErrMsg, ThisLoc )
               ENDIF
#if defined( USE_TIMERS )
               CALL GEOS_Timer_End ( "Dry deposition", RC )
#endif
               !### Debug
               IF ( prtDebug ) THEN 
                  CALL Debug_Msg( '### MAIN: a DO_DRYDEP' )
               ENDIF
            ENDIF
            !===========================================================
            !             ***** E M I S S I O N S *****
            !
            ! NOTE: For a complete description of how emissions from
            ! HEMCO are added into GEOS-Chem (and how they are mixed
            ! into the boundary layer), please see the wiki page:
            !
            ! http://wiki-geos-chem.org/Distributing_emissions_in_the_PBL
            !===========================================================
#if defined( USE_TIMERS )
            CALL GEOS_Timer_Start( "HEMCO", RC )
#endif         
            ! Is it time for emissions?
            TimeForEmis = ITS_TIME_FOR_EMIS()
            ! EMISSIONS_RUN will call HEMCO run phase 2. HEMCO run phase
            ! only calculates emissions. All data has been read to disk
            ! in phase 1 at the beginning of the time step. 
            ! (ckeller, 4/1/15)
            CALL Emissions_Run( am_I_Root,  Input_Opt,  State_Chm, 
     &                          State_Diag, State_Grid, State_Met,
     &                          TimeForEmis, 2,         RC )
            ! Trap potential errors
            IF ( RC /= GC_SUCCESS ) THEN
               ErrMsg = 
     &           'Error encountered in "Emissions_Run"! after drydep!'
               CALL Error_Stop( ErrMsg, ThisLoc )
            ENDIF
            !### Debug
            IF ( prtDebug ) THEN 
               CALL Debug_Msg( '### MAIN: a HEMCO PHASE 2' )
            ENDIF
# if defined( USE_TIMERS )
            CALL GEOS_Timer_End( "HEMCO", RC )
# endif
         ENDIF                //**_after exchange end _**

``` 
Does this change cause an error?  Thank you very much!
 huoxiao

**Comment by yantosca** [comment]
Thanks for writing.  I don't know much about the GEOS-Chem adjoint, because that is maintained separately from the forward model.

I believe that sequence is needed because you need to do dry deposition and emissions before mixing, because the emissions and drydep tendencies are applied into the boundary layer by the mixing scheme (VDIFF).  

If you want to change the order of operations, then you can.  My guess is that it would create some errors.  We would not be able to advise you further, as we do not maintain the adjoint.  We recommend that you contact the [Adjoint and Data Assimilation Working Group](http://wiki.seas.harvard.edu/geos-chem/index.php/GEOS-Chem_Adjoint) for more information.

**Comment by huoxiao-C** [comment]
Thank for your reply. Because in the GEOS-Chem 9.0.2 version, the emission subroutine is the last call. I guess according to the code's comments that the emission subroutine is the last call before 2015.
```
         ! Test for emission timestep
         ! Now always do emissions here, even for full-mixing
         ! (ckeller, 3/5/15)
```
I think the main reason is the non local pbl mixing,.
I ran for a month before the code was modified, I ran for a month after the code was run, and then I wanted to verify the results of the two to see if it would cause the wrong result.Because I ran the model lately, so the program is running now.
huoxiao
